### ISSUE TITLE
allow full pod spec (or maybe even statefulset spec)

### ISSUE THREAD

**Issue by redbaron** [zalando/postgres-operator#479]
Operator config already has init containers, side containers, pod service account name, environment vars configmap, resources and I am pretty sure more to come (for instance tolerations, DNSConfig, extra volumes/volume mounts).
 
It wont stop until all possible pod features will be reexported as arbitrary DIY  params , so maybe it is worth to allow full podSpec to be specified in operator config?  It then be used as a base and overlayed with any changes required by the runtime (like adding  environment variables with created credentials secrets or spilo configuration environment) when creating a statefulset.

**Comment by redbaron** [comment]
@Jan-M , I can work on implementation if idea is generally accepted and we discuss design at least ot the high level

**Comment by erthalion** [comment]
I agree, it sounds very appealing, and I myself missed this for testing purposes many times. But before jumping on it I guess we need to think through the possible use cases. As I see there are three possible options to configure pods:

* use only operator provided options (I guess, it's a most used option)
* use a part of a full podSpec and merge it with the rest of the options
* use a full podSpec to fully customize pods

And even taking a simplest possible approach (i.e. without a second case) I think it will require significant restructure of every code, which deals with configuration, to not make it looks like a hack. Another point is that it will make validation more complicated. At the end I think it makes sense to split this and tackle every question separately.

**Comment by redbaron** [comment]
@erthalion ,my proposal is to have full podSpec specified, then operator unconditionally overlays any necessary modifications on top of it.  If no podSpec is set, then operator works as it is now.

Then there is choice what to do with existing DIY customizations when podSpec is set, as you outlined there are really options:

- merge with podSpec
- reject CRD as one having conflicting requirements
- silently ignore with warning in logs

I prefer second option as it cleanest and least surprising: anyone who wants full control over pod have to give up any shortcuts and specify it in full.   Any attempt to reconcile and merge old confgs with podSpec would raise too many questions with too few good answers, such as:

- does side_containers append, replace or replace all but 1 containers in podSpec?
- which scalar values wins? `postgres.podSpec.containers[0].image` or `posgres.dockerImage` ?

**Comment by hedgss** [comment]
It is a really useful feature that covers a lot of cases when someone needs a custom configuration.
In my case, I need to be sure that Kubernetes pod uses storage in the same datacenter.
At this moment it is not possible to implement.
How long will this topic be in discussion state?
What is required to move this feature on the next stages?

**Comment by paalkr** [comment]
This is an excellent idea, and it would solve some important issues for us.
1) Be able to inject a custom security contex
2) Add a sidecar to tail the logs to stdout, using a pod shared volume mounted at /home/postgres/pgdata/pgroot/pg_log
3) Mount volumes with custom contend needed by some of our sidecar containers

I'm sure there are plenty of other use cases as well. Btw, KubeDB supports this feature
https://kubedb.com/docs/v0.13.0-rc.0/concepts/databases/postgres/#specpodtemplate

**Comment by frittentheke** [comment]
@paalkr your wish might just have come true: https://github.com/zalando/postgres-operator/pull/890

**Comment by fischerman** [comment]
@redbaron I really like your approach of having the PodSpec (i would even prefer StatefulsetSpec, maybe even ServiceSpec, etc...) as the basis. The operator takes this as input for validation and applying changes. This enables a very fine grained merging strategy but still provide some safety. Here is what I mean:

The operator would go through the provided template leaf by leaf. It sets fields which are left unspecified and complains about fields which are not compliant (e.g. everything that must stay in control of the operator or Patroni). For example, the operator looks for a container called `Postgres`. If it is not there, it creates one with reasonable default values. If it is there, it looks for container port with port number 5432. If it is not there it create the port. Otherwise it checks that the protocol is correctly set to `TCP`, otherwise rejects the manifest. This type of merging means, one can set any field (even partial containers) and the operator will attempt to patch it or complain. This behavior should be well documented, e.g. the name of the Spilo container must be stable, or how does it handle volumes, ....

This approach would only allow to specify the spec once. It becomes a little bit more complicated when considering that this approach is useful at the operator and cluster level.

As I was writing the above, I realized that this probably still needs some thinking.  I'm putting it out here anyways for inspiration.

**Comment by frittentheke** [comment]
@fischerman  maybe the way to go would not be merging the default PodSpec built into the operator with something that is configured, but actually making the PodSpec the operator uses just available for modification. Provided there is a set of variables that are expanded (just make the operator configuration as well as the CR fully available)  ... a Go template might work nicely.
Certainly this offers great flexibility but carries the risk that a config change makes even breaks the base functionality - but so can all settings even if they are "just" configuration flags or variables.

**Comment by cdmikechen** [comment]
@MPV @redbaron 
I have my own ideas on additional volumes.
If we need to expand new PVC, we need to use `volumeClaimTemplates`. However,`StatefulSet` currently has restrictions on`volumeClaimTemplates`, so we can't add a new pvcs to this. Therefore, this is a contradiction, which can only be solved by deleting the existing `StatefulSet` and rebuilding. 
But in this way, the database will not be available for a period of time. Theoretically, this is a necessary step for the operator. It also needs to change a lot of code and test.

**Comment by FxKu** [comment]
We decided against allowing full podSpec. Yes it can be a race with supporting future K8s version but it gives us better control over what our teams can and can not use.
